### ISSUE TITLE
Properly support and test nested namespaces

### ISSUE THREAD

**Issue by Manishearth** [rust-diplomat/diplomat#591]
Nested C++ namespaces used to work, but they weren't important to ICU4X so we weren't testing it. They've been broken by https://github.com/rust-diplomat/diplomat/pull/572. 

Ideally we support nested namespaces for all backends. We can continue to use `foo::bar` syntax, potentially also supporting `foo.bar`, and having the backends parse them into the right thing.

There's a potentially annoying difference between C++ and JS/Dart here: in C++ you basically _need_ a namespace, whereas in JS/Dart you can kind of get by without one, and using more modules is just a code organization thing.

A way to talk about actual namespaces vs the "library name" seems useful. Could be a global config. See https://github.com/rust-diplomat/diplomat/issues/589 for more info.

**Comment by robertbastian** [comment]
> They've been broken by https://github.com/rust-diplomat/diplomat/pull/572.

Oh I didn't know, if there'd been a test I would've supported them.

> There's a potentially annoying difference between C++ and JS/Dart here: in C++ you basically need a namespace, whereas in JS/Dart you can kind of get by without one, and using more modules is just a code organization thing.

Well, if we have a (required) `library_name` attribute, that can be the top-level C++ namespace, and the special thing in Dart/Js/Kotlin/etc.

**Comment by Manishearth** [comment]
Yeah, that's my bad. @sffc was also surprised during a discussion where he was confused as to the constraints I was operating under until it turned out that he thought that nested namespaces did work.

But it's not a big deal, we can fix it later, and we can track that with this issue.

> Well, if we have a (required) `library_name` attribute, that can be the top-level C++ namespace, and the special thing in Dart/Js/Kotlin/etc.

Yeah, though I'm wondering if it should actually be a config rather than an attribute. Dart/JS presumably cannot generate multiple libraries in one bundle.

Kotlin and C++ can.
